src/main/java/bio/terra/service/snapshot/SnapshotDao.java

```diff
@@ -513,6 +514,31 @@ class RawSourceData {
     return snapshotSources;
   }
 
+  /**
+   * @param permissions RAS dbGaP permissions held by the caller (derived from the caller's linked
+   *     RAS passports)
+   * @return snapshot UUIDs accessible under the permissions
+   */
+  @Transactional(
+      propagation = Propagation.REQUIRED,
+      isolation = Isolation.SERIALIZABLE,
+      readOnly = true)
+  public List<UUID> getAccessibleSnapshots(List<RASDbgapPermissions> permissions) {
+    String sql =
+        "SELECT snapshot.id FROM snapshot "
+            + "JOIN snapshot_source ON snapshot.id = snapshot_source.snapshot_id "
+            + "JOIN dataset ON dataset.id = snapshot_source.dataset_id "
+            + "WHERE (snapshot.consent_code, dataset.phs_id) IN (:criteria)";
```

Comment:
We should only return snapshots whose permissions fully match one of our `RASDbgapPermissions` objects, no partial matches permitted.